### PR TITLE
[ENH] Data Loader for M5 dataset

### LINKED ISSUE #3296: [ENH] Data Loader for M5 Dataset
**Is your feature request related to a problem? Please describe.**

At the moment, there is no data loader for M5 dataset. Addition of this dataset will help greatly enhance our bench marking ability. The M5 dataset gives historical sales data and asked users to predict sales for two 28-day time periods. The dataset contains the following files:
* calendar.csv - Contains information about the dates on which the products are sold.
* sales_train_validation.csv - Contains the historical daily unit sales data per product and store [d_1 - d_1913]
* sample_submission.csv - The correct format for submissions.
* sell_prices.csv - Contains information about the price of the products sold per store and date.
* sales_train_evaluation.csv - Includes sales [d_1 - d_1941] (labels used for the Public leaderboard)

Original data can be found at [Kaggle](https://www.kaggle.com/competitions/m5-forecasting-accuracy/data)

**Describe the solution you'd like**
A dataset loader for M5 dataset. It'll work as follows:
`load_M5()`
The result will be (a) Dataframe(s) in memory or a Callable that fetches data as the forecasting goes on. The dataset loader will prepare the dataset so that it's ready for use.

**Describe alternatives you've considered**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by SaiRevanth25**
#### Reference Issues/PRs
 Fixes #3296 

#### What does this implement/fix? Explain your changes.
 
`load_m5` function downloads the m5 dataset if not downloaded, preprocess it and make it `sktime` compatible.

` _reduce_memory_usage ` function reduces the memory usage of the datasets by downcasting numerical 
    columns to more memory-efficient types and converting object columns to the category data type.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by SaiRevanth25** [comment]
The failures are not regarding to this PR, it seems to be from `test_load_solar`

**Review by fkiraly — APPROVED** [review]
Great, looks good!

Could you also add this to the API reference for data loaders? In `datasets.rst`

It would also be great if we could somehow test it, I understand that this is relatively large, but could there be a way to add sth like a test run or a partial run and test that?

**Comment by SaiRevanth25** [comment]
> Could you also add this to the API reference for data loaders? In datasets.rst

sure!!

> but could there be a way to add sth like a test run or a partial run and test that?

We could add another parameter, let's say `test`, to the function to specify a smaller sample size for the dataset and use it when running the tests.

**Comment by SaiRevanth25** [comment]
the `loaded_dataset` is of type `pd_multiindex_hier`, idk why the tests are failing!!

**Comment by fkiraly** [comment]
This now seems to work?

**Comment by SaiRevanth25** [comment]
yes!!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks, this is great!

I have one question, and would kindly be requesting improvements to the docstring.

* preamble should say that it is loaded in an `sktime` in-memory format, and which
* I would also suggest to provide some minimal description of the dataset, e.g., number of rows, levels, etc, that the user should expect.
* `extract_path` precise format needs to be explained, same for `folder_path`
* `include_events` - unclear what this does. What does this do?
* `test` - not concrete. It should be said very explicitly that this loads a smaller part for testing. This should not be used in vanilla usage, but a developer might find it useful for testing.


Question: why do we need to pass `extract_path` and `folder_path` both? I am confused how I would use it.

**Comment by SaiRevanth25** [comment]
> Question: why do we need to pass extract_path and folder_path both? I am confused how I would use it.

- The `extract_path` will download and extract the data. It creates a folder named `data`, and within it, a subfolder `m5-forecasting-accuracy`, where the 5 .csv files will be downloaded. The data is then preprocessed after downloading.
- The `folder_path` is provided if the data has already been downloaded. It is the path to the folder that contains the .csv files.it avoid the downloading step and directly preprocesses the data.

**Comment by fkiraly** [comment]
>     * The `extract_path` will download and extract the data. It creates a folder named `data`, and within it, a subfolder `m5-forecasting-accuracy`, where the 5 .csv files will be downloaded. The data is then preprocessed after downloading.
> 
>     * The `folder_path` is provided if the data has already been downloaded. It is the path to the folder that contains the .csv files.it avoid the downloading step and directly preprocesses the data.

These sound redundant to me, since at most one of the two will apply in a given workflow, no?

I would provide only one path, and:

* if no file exists in location, download.
* if file exists, assume it is the downloaded data, and extract. Possibly check if it is the right file.

**Comment by SaiRevanth25** [comment]
- updated the doc strings.
- made changes w.r.t `extract_path` and removed the `folder_path` parameter

File: sktime/datasets/_single_problem_loaders.py [hunk 1]
```diff
@@ -1438,3 +1439,233 @@ def load_forecastingdata(
...
+    """Fetch M5 dataset from https://zenodo.org/records/12636070 .
+
+    Downloads and extracts dataset if not already downloaded. Fetched dataset is
+    in the standard .csv format and loaded into an sktime-compatible in-memory
+    format (`pd_multiindex_hier`).
+
+    Parameters
+    ----------
+    extract_path : str, optional (default=None)
+        The path to extract the data to. If the provided path doesn't
```
Comment (fkiraly) [inline sktime/datasets/_single_problem_loaders.py:1459]: * I would be more specific here on what it means to "already contain the data" - a path with all files present, from a previous run of the function?
* be more specific about the kind of string expected here. If relative, what is it going to be relative to?

File: sktime/datasets/_single_problem_loaders.py [hunk 2]
```diff
@@ -1438,3 +1439,233 @@ def load_forecastingdata(
...
+    extract_path=None,
+    folder_path=None,
+    include_events=False,
+    test=False,
+):
+    """Fetch M5 dataset from https://zenodo.org/records/12636070 .
+
+    Downloads and extracts dataset if not already downloaded. Fetched dataset is
+    in the standard .csv format and loaded into an sktime-compatible in-memory
+    format (`pd_multiindex_hier`).
```
Comment (fkiraly) [inline sktime/datasets/_single_problem_loaders.py:1454]: point to the "notes" section for further details on the data.

File: sktime/datasets/_single_problem_loaders.py [hunk 3]
```diff
@@ -1438,3 +1439,233 @@ def load_forecastingdata(
...
+    Dataset Description
+    --------------------
+    - **Number of Rows**: Approximately 58 million rows (for the full dataset).
+    - **Number of Columns**: Varies based on `include_events` parameter.
+      - Without events: 9 columns.
+      - With events: 13 columns.
+
+    Notes
+    -----
+    The dataset consists of three main files:
```
Comment (fkiraly) [inline sktime/datasets/_single_problem_loaders.py:1492]: something does not seem right. Should these not be loaded into three data frames?

File: sktime/datasets/_single_problem_loaders.py [hunk 4]
```diff
@@ -1438,3 +1439,233 @@ def load_forecastingdata(
...
+                    "sales",
+                    "date",
+                    "wm_yr_wk",
+                    "wday",
+                    "month",
+                    "year",
+                ]
+            ]
+
+        df4 = df3.merge(sp, on=["store_id", "item_id", "wm_yr_wk"], how="left")
```
Comment (fkiraly) [inline sktime/datasets/_single_problem_loaders.py:1637]: oh, I see, you are merging

**Review by fkiraly — CHANGES_REQUESTED** [review]
I left some comments related to the docstring. These are not blocking.

My key comment is (after trying this out) is that the loader merges the three data tables, and the user does not have the option to not merge them.

I would suggest an interface where the user can choose whether they want them merged, or returned individually, e.g., `merged:bool=False`

**Comment by SaiRevanth25** [comment]
> I would suggest an interface where the user can choose whether they want them merged, or returned individually, e.g., merged:bool=False

- if not merged, the dataframe will not be of type `pd_multiindex_hier` would that be fine? will it be sktime compatible?

- And in the merged dataframe, should we convert the the state_id, store_id, cat_id, dept_id into cat codes/ labelencode them?

**Comment by fkiraly** [comment]
> if not merged, the dataframe will not be of type `pd_multiindex_hier`

Can you kindly explain that? Would it not be sales at multiple hierarchies, still? At least the `sales_train_validation.csv`?

**Comment by fkiraly** [comment]
> should we convert the columns state_id, store_id, cat_id, dept_id into cat codes/ labelencode before making them indexes?

I would not manipulate them, and stay as close to the original data as possible.

File: sktime/datasets/_single_problem_loaders.py (hunk 1 above)
Comment (SaiRevanth25) [reply sktime/datasets/_single_problem_loaders.py:1459]:
```
extract_path : str, optional (default=None)

If provided, the path should use the appropriate path separators for the operating system.(e.g., forward slashes '/' for Unix-based systems, backslashes '\\' for Windows).
- If `extract_path` is provided:
    - Check if the required files (.csv) are present at the given `extract_path`.
    - If files are not found, check if the directory "m5-forecasting-accuracy" exists within the `extract_path`. Useful when the function has already run previously with the same path.
    - If the directory does not exist, download and extract the data  into "m5-forecasting-accuracy" folder in the `extract_path`.
    - If the directory exists, takes the path to the existing directory.
- if `extract_path` is None:
    - Check if the directory "m5-forecasting-accuracy" exists within the module level.
    - If the directory exists, takes path to current directory. Useful when the function has already run previously without any path.
    - If the directory does not exist, download and extract the data into "m5-forecasting-accuracy" folder at the module level.
```

Please let me know if i have to add anything or missed something!!

**Review by fkiraly — APPROVED** [review]

**Pull request merged**